python/lsst/ts/ess/csc/data_client/young_32400_weather_station_data_client.py

```diff
@@ -0,0 +1,713 @@
+                    self.cached_rain_tip_timestamp = timestamp
+                    rain_rate_mm_per_hr = (
+                        rain_tip_delta
+                        * self.config.scale_rain_rate
+                        * SECONDS_PER_HOUR
+                        / rain_tip_dt
+                    )
+                    # TODO DM-38119: remove the int cast.
+                    await self.topics.tel_rainRate.set_write(
+                        rainRate=int(round(rain_rate_mm_per_hr))
+                    )
+
+    def restart_rain_stopped_timer(self) -> None:
+        """Start or restart the "rain stopped" timer."""
+        self.rain_stopped_timer_task.cancel()
+        self.rain_stopped_timer_task = asyncio.create_task(self.rain_stopped_timer())
+
+    async def run(self) -> None:
+        self.read_loop_task.cancel()
+        self.read_loop_task = asyncio.create_task(self.read_loop())
+
+    async def read_loop(self) -> None:
+        """Read raw data from the weather station.
+
+        The format is as described by by DATA_REGEX.
+        """
+        simulated_data_iter = iter(self.simulated_raw_data)
+        try:
+            while self.connected or self.simulation_mode != 0:
+                if self.simulation_mode == 0:
+                    assert self.client is not None  # make mypy happy
+                    read_bytes = await asyncio.wait_for(
+                        self.client.read_until(TERMINATOR),
+                        timeout=self.config.read_timeout,
+                    )
+                    data = read_bytes.decode().strip()
+                else:
+                    await asyncio.sleep(self.simulation_interval)
+                    data = next(simulated_data_iter)
+
+                if not data:
+                    continue
+                match = DATA_REGEX.fullmatch(data)
+                if match is None:
+                    self.log.warning(f"Ignoring {data=}: could not parse the data")
+                    continue
+                # Convert raw data values from str to int.
+                raw_data_dict = {
+                    key: int(value) for key, value in match.groupdict().items()
+                }
+                try:
+                    await self.handle_data(**raw_data_dict)
+                except Exception as e:
+                    self.log.exception(f"Failed to handle {data=}: {e!r}")
+        except StopIteration:
+            self.log.info("read loop ends: out of simulated raw data")
+        except Exception as e:
+            self.log.exception(f"read loop failed: {e!r}")
+            raise
+
+    async def rain_stopped_timer(self) -> None:
+        """Wait for the configured time, then report that rain has stopped.
+
+        Intended to be run by restart_rain_stopped_timer.
+        """
+        await asyncio.sleep(self.config.rain_report_interval)
+        await self.topics.evt_precipitation.set_write(raining=False)
+
+
+@dataclasses.dataclass
+class Young32400RawDataGenerator:
```

Comment:
This should be a subclass of the ts_ess_common BaseSensor and added to that package.

Reply:
I guess. This is that weird regime where "a sensor" is really a whole set of sensors. But the API looks OK.